// endomondo-api-handler/dist/models/Workout.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _defineProperty = _interopDefault(require('@babel/runtime/helpers/defineProperty'));
var luxon = require('luxon');
var mathjs = require('mathjs');
var fitnessModels = require('fitness-models');
var gpxBuilder = require('gpx-builder');

const RUNNING = 0;
const CYCLING_TRANSPORT = 1;
const CYCLING_SPORT = 2;
const MOUNTAIN_BIKINGS = 3;
const SKATING = 4;
const ROLLER_SKIING = 5;
const SKIING_CROSS_COUNTRY = 6;
const SKIING_DOWNHILL = 7;
const SNOWBOARDING = 8;
const KAYAKING = 9;
const KITE_SURFING = 10;
const ROWING = 11;
const SAILING = 12;
const WINDSURFING = 13;
const FINTESS_WALKING = 14;
const GOLFING = 15;
const HIKING = 16;
const ORIENTEERING = 17;
const WALKING = 18;
const RIDING = 19;
const SWIMMING = 20;
const CYCLING_INDOOR = 21;
const OTHER = 22;
const AEROBICS = 23;
const BADMINTON = 24;
const BASEBALL = 25;
const BASKETBALL = 26;
const BOXING = 27;
const CLIMBING_STAIRS = 28;
const CRICKET = 29;
const ELLIPTICAL_TRAINING = 30;
const DANCING = 31;
const FENCING = 32;
const FOOTBALL_AMERICAN = 33;
const FOOTBALL_RUGBY = 34;
const FOOTBALL_SOCCER = 35;
const HANDBALL = 36;
const HOCKEY = 37;
const PILATES = 38;
const POLO = 39;
const SCUBA_DIVING = 40;
const SQUASH = 41;
const TABLE_TENIS = 42;
const TENNIS = 43;
const VOLEYBALL_BEACH = 44;
const VOLEYBALL_INDOOR = 45;
const WEIGHT_TRAINING = 46;
const YOGA = 47;
const MARTINAL_ARTS = 48;
const GYMNASTICS = 49;
const STEP_COUNTER = 50;
const CIRKUIT_TRAINING = 87;
const RUNNING_TREADMILL = 88;
const SKATEBOARDING = 89;
const SURFING = 90;
const SNOWSHOEING = 91;
const WHEELCHAIR = 92;
const CLIMBING = 93;
const WALKING_TREADMILL = 94;
const KICK_SCOOTER = 95;
const STAND_UP_PADDLING = 96;
const TRAIL_RUNNING = 97;
const ROWING_INDOORS = 98;
const FLOORBALL = 99;
const ICE_SKATING = 100;
const SKI_TOURING = 101;
const ROPE_JUMPING = 102;
const STRETCHING = 103;
const CANICROSS = 104;
const PADDLE_TENNIS = 105;
const PARAGLIDING = 106;

var SPORT = ({
    __proto__: null,
    RUNNING: RUNNING,
    CYCLING_TRANSPORT: CYCLING_TRANSPORT,
    CYCLING_SPORT: CYCLING_SPORT,
    MOUNTAIN_BIKINGS: MOUNTAIN_BIKINGS,
    SKATING: SKATING,
    ROLLER_SKIING: ROLLER_SKIING,
    SKIING_CROSS_COUNTRY: SKIING_CROSS_COUNTRY,
    SKIING_DOWNHILL: SKIING_DOWNHILL,
    SNOWBOARDING: SNOWBOARDING,
    KAYAKING: KAYAKING,
    KITE_SURFING: KITE_SURFING,
    ROWING: ROWING,
    SAILING: SAILING,
    WINDSURFING: WINDSURFING,
    FINTESS_WALKING: FINTESS_WALKING,
    GOLFING: GOLFING,
    HIKING: HIKING,
    ORIENTEERING: ORIENTEERING,
    WALKING: WALKING,
    RIDING: RIDING,
    SWIMMING: SWIMMING,
    CYCLING_INDOOR: CYCLING_INDOOR,
    OTHER: OTHER,
    AEROBICS: AEROBICS,
    BADMINTON: BADMINTON,
    BASEBALL: BASEBALL,
    BASKETBALL: BASKETBALL,
    BOXING: BOXING,
    CLIMBING_STAIRS: CLIMBING_STAIRS,
    CRICKET: CRICKET,
    ELLIPTICAL_TRAINING: ELLIPTICAL_TRAINING,
    DANCING: DANCING,
    FENCING: FENCING,
    FOOTBALL_AMERICAN: FOOTBALL_AMERICAN,
    FOOTBALL_RUGBY: FOOTBALL_RUGBY,
    FOOTBALL_SOCCER: FOOTBALL_SOCCER,
    HANDBALL: HANDBALL,
    HOCKEY: HOCKEY,
    PILATES: PILATES,
    POLO: POLO,
    SCUBA_DIVING: SCUBA_DIVING,
    SQUASH: SQUASH,
    TABLE_TENIS: TABLE_TENIS,
    TENNIS: TENNIS,
    VOLEYBALL_BEACH: VOLEYBALL_BEACH,
    VOLEYBALL_INDOOR: VOLEYBALL_INDOOR,
    WEIGHT_TRAINING: WEIGHT_TRAINING,
    YOGA: YOGA,
    MARTINAL_ARTS: MARTINAL_ARTS,
    GYMNASTICS: GYMNASTICS,
    STEP_COUNTER: STEP_COUNTER,
    CIRKUIT_TRAINING: CIRKUIT_TRAINING,
    RUNNING_TREADMILL: RUNNING_TREADMILL,
    SKATEBOARDING: SKATEBOARDING,
    SURFING: SURFING,
    SNOWSHOEING: SNOWSHOEING,
    WHEELCHAIR: WHEELCHAIR,
    CLIMBING: CLIMBING,
    WALKING_TREADMILL: WALKING_TREADMILL,
    KICK_SCOOTER: KICK_SCOOTER,
    STAND_UP_PADDLING: STAND_UP_PADDLING,
    TRAIL_RUNNING: TRAIL_RUNNING,
    ROWING_INDOORS: ROWING_INDOORS,
    FLOORBALL: FLOORBALL,
    ICE_SKATING: ICE_SKATING,
    SKI_TOURING: SKI_TOURING,
    ROPE_JUMPING: ROPE_JUMPING,
    STRETCHING: STRETCHING,
    CANICROSS: CANICROSS,
    PADDLE_TENNIS: PADDLE_TENNIS,
    PARAGLIDING: PARAGLIDING
});

var LIST_OF_SPORT_NAMES = {
  [RUNNING]: 'Running',
  [CYCLING_TRANSPORT]: 'Cycling, transport',
  [CYCLING_SPORT]: 'Cycling, sport',
  [MOUNTAIN_BIKINGS]: 'Mountain biking',
  [SKATING]: 'Skating',
  [ROLLER_SKIING]: 'Roller skiing',
  [SKIING_CROSS_COUNTRY]: 'Skiing, cross country',
  [SKIING_DOWNHILL]: 'Skiing, downhill',
  [SNOWBOARDING]: 'Snowboarding',
  [KAYAKING]: 'Kayaking',
  [KITE_SURFING]: 'Kite surfing',
  [ROWING]: 'Rowing',
  [SAILING]: 'Sailing',
  [WINDSURFING]: 'Windsurfing',
  [FINTESS_WALKING]: 'Fitness walking',
  [GOLFING]: 'Golfing',
  [HIKING]: 'Hiking',
  [ORIENTEERING]: 'Orienteering',
  [WALKING]: 'Walking',
  [RIDING]: 'Riding',
  [SWIMMING]: 'Swimming',
  [CYCLING_INDOOR]: 'Cycling, Indoor',
  [OTHER]: 'Other',
  [AEROBICS]: 'Aerobics',
  [BADMINTON]: 'Badminton',
  [BASEBALL]: 'Baseball',
  [BASKETBALL]: 'Basketball',
  [BOXING]: 'Boxing',
  [CLIMBING_STAIRS]: 'Climbing stairs',
  [CRICKET]: 'Cricket',
  [ELLIPTICAL_TRAINING]: 'Elliptical training',
  [DANCING]: 'Dancing',
  [FENCING]: 'Fencing',
  [FOOTBALL_AMERICAN]: 'Football, American',
  [FOOTBALL_RUGBY]: 'Football, rugby',
  [FOOTBALL_SOCCER]: 'Football, soccer',
  [HANDBALL]: 'Handball',
  [HOCKEY]: 'Hockey',
  [PILATES]: 'Pilates',
  [POLO]: 'Polo',
  [SCUBA_DIVING]: 'Scuba diving',
  [SQUASH]: 'Squash',
  [TABLE_TENIS]: 'Table tennis',
  [TENNIS]: 'Tennis',
  [VOLEYBALL_BEACH]: 'Volleyball, beach',
  [VOLEYBALL_INDOOR]: 'Volleyball, indoor',
  [WEIGHT_TRAINING]: 'Weight training',
  [YOGA]: 'Yoga',
  [MARTINAL_ARTS]: 'Martial arts',
  [GYMNASTICS]: 'Gymnastics',
  [STEP_COUNTER]: 'Step counter',
  [CIRKUIT_TRAINING]: 'Circuit Training',
  [SKATEBOARDING]: 'Skateboarding',
  [CLIMBING]: 'Climbing',
  [KICK_SCOOTER]: 'Kick scooter',
  [CANICROSS]: 'Canicross',
  [FLOORBALL]: 'Floorball',
  [ICE_SKATING]: 'Ice skating',
  [RUNNING_TREADMILL]: 'Running (Treadmill)',
  [SURFING]: 'Surfing',
  [SNOWSHOEING]: 'Showshoeing',
  [WHEELCHAIR]: 'Wheelchair',
  [WALKING_TREADMILL]: 'Walking (Treadmill)',
  [STAND_UP_PADDLING]: 'Stand up paddling',
  [TRAIL_RUNNING]: 'Trail running',
  [ROWING_INDOORS]: 'Rowing (indoors)',
  [SKI_TOURING]: 'Ski touring',
  [ROPE_JUMPING]: 'Rope jumping',
  [STRETCHING]: 'Stretching',
  [PADDLE_TENNIS]: 'Paddle tennis',
  [PARAGLIDING]: 'Paragliding'
};

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }
class Point extends fitnessModels.Point {
  constructor(options) {
    super(options);

    _defineProperty(this, "instruction", void 0);

    this.instruction = options.instruction;
  }

  static fromApi(point, timezone) {
    const {
      distance,
      altitude,
      sensor_data,
      time
    } = point;
    return new Point(_objectSpread({
      time: time ? luxon.DateTime.fromISO(time, {
        zone: timezone
      }) : undefined,
      instruction: point.instruction,
      latitude: point.latitude,
      longitude: point.longitude,
      duration: luxon.Duration.fromObject({
        seconds: point.duration
      }),
      distance: distance != null ? mathjs.unit(distance, 'km') : undefined,
      altitude: altitude != null ? mathjs.unit(altitude, 'm') : undefined
    }, sensor_data ? {
      speed: sensor_data.speed != null ? mathjs.unit(sensor_data.speed, 'km/h') : undefined,
      hr: sensor_data.heart_rate,
      cadence: sensor_data.cadence
    } : {}));
  } // eslint-disable-next-line complexity


  static get(time, latitude, longitude, {
    instruction,
    distance,
    duration,
    speed,
    altitude,
    cadence,
    hr
  } = {}) {
    return new Point(_objectSpread({
      time: time instanceof luxon.DateTime ? time : luxon.DateTime.fromISO(time, {
        setZone: true
      }),
      latitude,
      longitude,
      hr,
      instruction,
      cadence,
      distance: typeof distance === 'number' ? mathjs.unit(distance, 'km') : distance,
      altitude: typeof altitude === 'number' ? mathjs.unit(altitude, 'm') : altitude,
      speed: typeof speed === 'number' ? mathjs.unit(speed, 'km/h') : speed
    }, duration instanceof luxon.Duration ? {
      duration
    } : {}, {}, typeof duration === 'number' ? {
      duration: luxon.Duration.fromObject({
        seconds: duration
      })
    } : {}, {}, !(duration instanceof luxon.Duration) && typeof duration === 'object' ? {
      duration: luxon.Duration.fromObject(duration)
    } : {}));
  }

  clone(extension = {}) {
    return new Point(_objectSpread({}, this.toObject(), {}, extension));
  }

  getInstruction() {
    return this.instruction;
  }

  setInstruction(instruction) {
    return this.clone({
      instruction
    });
  }

  setTime(time) {
    return this.clone({
      time
    });
  }

  setLatitude(latitude) {
    return this.clone({
      latitude
    });
  }

  setLongitude(longitude) {
    return this.clone({
      longitude
    });
  }

  setAltitude(altitude) {
    return this.clone({
      altitude
    });
  }

  setDistance(distance) {
    return this.clone({
      distance
    });
  }

  setSpeed(speed) {
    return this.clone({
      speed
    });
  }

  setHeartRate(hr) {
    return this.clone({
      hr
    });
  }

  setCadence(cadence) {
    return this.clone({
      cadence
    });
  }

  setDuration(duration) {
    return this.clone({
      duration
    });
  }

  toObject() {
    return _objectSpread({}, super.toObject(), {
      instruction: this.getInstruction()
    });
  }

  toString() {
    const distance = this.getDistance();
    const altitude = this.getAltitude();
    const speed = this.getSpeed();
    const time = this.getTime();
    return [time != null ? time.toUTC().toFormat('yyyy-MM-dd HH:mm:ss \'UTC\'') : undefined, this.getInstruction(), this.getLatitude(), this.getLongitude(), distance != null ? distance.toNumber('km') : undefined, speed != null ? speed.toNumber('km/h') : undefined, altitude != null ? altitude.toNumber('m') : undefined, this.getHeartRate(), this.getCadence(), ''].map(item => {
      return item == null ? '' : item;
    }).join(';');
  }

}

function ownKeys$1(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread$1(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys$1(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys$1(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }
const {
  Point: Point$1,
  Metadata,
  Person,
  Link,
  Track,
  Segment
} = gpxBuilder.GarminBuilder.MODELS; // @ts-ignore

function convertPoints(points) {
  return points.map(point => {
    return point.toObject();
  }).map(point => {
    const {
      altitude,
      cadence,
      hr,
      latitude,
      longitude,
      speed,
      time
    } = point;

    if (!latitude || !longitude) {
      return null;
    }

    return new Point$1(latitude, longitude, {
      time: time ? time.toJSDate() : undefined,
      hr,
      cad: cadence,
      ele: altitude ? altitude.toNumber('m') : undefined,
      speed: speed ? speed.toNumber('m/s') : undefined
    });
  }).filter(item => item !== null);
}

var workoutGPXExporter = (workout => {
  const workoutId = workout.getId();
  const source = workout.getSource();
  const authorId = source && source.author ? source.author.id : null;
  const authorName = source && source.author ? source.author.name : null;
  const builder = new gpxBuilder.GarminBuilder();
  builder.setMetadata(new Metadata(_objectSpread$1({}, authorName ? {
    author: new Person({
      name: authorName
    })
  } : {}, {
    link: new Link('http://www.endomondo.com', {
      text: 'Endomondo'
    }),
    time: workout.getStart().toJSDate()
  })));
  builder.setTracks([new Track([new Segment(convertPoints(workout.getPoints()))], _objectSpread$1({
    src: 'http://www.endomondo.com/'
  }, workoutId && authorId ? {
    link: new Link("https://www.endomondo.com/users/" + authorId + "/workouts/" + workoutId, {
      text: 'endomondo'
    })
  } : {}, {
    type: workout.getSportName()
  }))]);
  return gpxBuilder.buildGPX(builder.toObject());
});

function ownKeys$2(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread$2(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys$2(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys$2(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }
class Workout extends fitnessModels.Workout {
  constructor(options) {
    super(options);

    _defineProperty(this, "id", void 0);

    _defineProperty(this, "typeId", void 0);

    _defineProperty(this, "points", void 0);

    _defineProperty(this, "hashtags", void 0);

    _defineProperty(this, "source", void 0);

    _defineProperty(this, "message", void 0);

    this.typeId = options.typeId;
    this.points = options.points || [];
    this.privacy = options.workoutPrivacy;
    this.hashtags = options.hashtags || [];
    this.id = options.id;
    this.source = options.source;
    this.message = options.message;
    this.isRace = this.hasHashtag('race');
    this.isCommute = this.hasHashtag('work');
  }

  static fromApi(workout) {
    const {
      points,
      distance
    } = workout;
    const start = luxon.DateTime.fromISO(workout.local_start_time, {
      setZone: true
    });
    return new Workout(_objectSpread$2({
      start,
      typeId: workout.sport,
      duration: luxon.Duration.fromObject({
        seconds: workout.duration
      }),
      source: workout,
      points: points && points.points ? points.points.map(point => {
        return Point.fromApi(point, start.toFormat('z'));
      }) : [],
      ascent: workout.ascent ? mathjs.unit(workout.ascent, 'm') : undefined,
      descent: workout.descent ? mathjs.unit(workout.descent, 'm') : undefined,
      calories: workout.calories,
      notes: workout.message,
      mapPrivacy: workout.show_map,
      workoutPrivacy: workout.show_workout,
      id: workout.id,
      hashtags: workout.hashtags,
      avgHeartRate: workout.heart_rate_avg,
      maxHeartRate: workout.heart_rate_max,
      title: workout.title
    }, distance ? {
      distance: mathjs.unit(distance, 'km')
    } : {}));
  } // eslint-disable-next-line max-params


  static get(typeId, start, duration, distance, points = [], options = {}) {
    return new Workout(_objectSpread$2({}, options, {
      start,
      duration,
      distance,
      points,
      typeId,
      id: undefined,
      source: undefined
    }));
  }

  clone(extension) {
    return new Workout(_objectSpread$2({}, this.toObject(), {}, extension));
  }

  getId() {
    return this.id;
  }

  setId(id) {
    return this.clone({
      id
    });
  }

  getTypeId() {
    return this.typeId;
  }

  getSportName() {
    return Workout.SPORT_NAMES[this.getTypeId()];
  }

  getPoints() {
    return this.points;
  }

  getWorkoutPrivacy() {
    return this.privacy;
  }

  setWorkoutPrivacy(workoutPrivacy) {
    return this.clone({
      workoutPrivacy
    });
  }

  getMessage() {
    return this.message;
  }

  setMessage(message) {
    return this.clone({
      message
    });
  }

  setHashtags(hashtags) {
    return this.clone({
      hashtags
    });
  }

  addHashtags(hashtags) {
    return this.clone({
      hashtags: [...this.getHashtags(), ...hashtags]
    });
  }

  addHashtag(hashtag) {
    return this.addHashtags([hashtag]);
  }

  removeHashtag(hashtag) {
    return this.removeHashtags([hashtag]);
  }

  removeHashtags(hashtags) {
    return this.clone({
      hashtags: hashtags.filter(hashtag => hashtags.includes(hashtag))
    });
  }

  getSource() {
    return this.source;
  }

  toGpx() {
    return workoutGPXExporter(this);
  }

  hasGPSData() {
    return this.points.length > 0;
  }

  setTypeId(typeId) {
    return this.clone({
      typeId
    });
  }

  setStart(start) {
    return this.clone({
      start
    });
  }

  setDuration(duration) {
    return this.clone({
      duration
    });
  }

  setDistance(distance) {
    return this.clone({
      distance
    });
  }

  setPoints(points) {
    return this.clone({
      points
    });
  }

  setCalories(calories) {
    return this.clone({
      calories
    });
  }

  setNotes(notes) {
    return this.clone({
      notes
    });
  }

  setAvgHeartRate(avgHeartRate) {
    return this.clone({
      avgHeartRate
    });
  }

  setMaxHeartRate(maxHeartRate) {
    return this.clone({
      maxHeartRate
    });
  }

  setTitle(title) {
    return this.clone({
      title
    });
  }

  setAscent(ascent) {
    return this.clone({
      ascent
    });
  }

  setDescent(descent) {
    return this.clone({
      descent
    });
  }

  toObject() {
    return _objectSpread$2({}, super.toObject(), {
      typeId: this.typeId,
      points: this.points,
      mapPrivacy: this.mapPrivacy,
      workoutPrivacy: this.privacy,
      hashtags: this.hashtags,
      id: this.id,
      source: this.source,
      message: this.message
    });
  }

}

_defineProperty(Workout, "SPORT_NAMES", LIST_OF_SPORT_NAMES);

_defineProperty(Workout, "SPORT", SPORT);

module.exports = Workout;
