// endomondo-api-handler/dist/constants/index.esm.js
const RUNNING = 0;
const CYCLING_TRANSPORT = 1;
const CYCLING_SPORT = 2;
const MOUNTAIN_BIKINGS = 3;
const SKATING = 4;
const ROLLER_SKIING = 5;
const SKIING_CROSS_COUNTRY = 6;
const SKIING_DOWNHILL = 7;
const SNOWBOARDING = 8;
const KAYAKING = 9;
const KITE_SURFING = 10;
const ROWING = 11;
const SAILING = 12;
const WINDSURFING = 13;
const FINTESS_WALKING = 14;
const GOLFING = 15;
const HIKING = 16;
const ORIENTEERING = 17;
const WALKING = 18;
const RIDING = 19;
const SWIMMING = 20;
const CYCLING_INDOOR = 21;
const OTHER = 22;
const AEROBICS = 23;
const BADMINTON = 24;
const BASEBALL = 25;
const BASKETBALL = 26;
const BOXING = 27;
const CLIMBING_STAIRS = 28;
const CRICKET = 29;
const ELLIPTICAL_TRAINING = 30;
const DANCING = 31;
const FENCING = 32;
const FOOTBALL_AMERICAN = 33;
const FOOTBALL_RUGBY = 34;
const FOOTBALL_SOCCER = 35;
const HANDBALL = 36;
const HOCKEY = 37;
const PILATES = 38;
const POLO = 39;
const SCUBA_DIVING = 40;
const SQUASH = 41;
const TABLE_TENIS = 42;
const TENNIS = 43;
const VOLEYBALL_BEACH = 44;
const VOLEYBALL_INDOOR = 45;
const WEIGHT_TRAINING = 46;
const YOGA = 47;
const MARTINAL_ARTS = 48;
const GYMNASTICS = 49;
const STEP_COUNTER = 50;
const CIRKUIT_TRAINING = 87;
const RUNNING_TREADMILL = 88;
const SKATEBOARDING = 89;
const SURFING = 90;
const SNOWSHOEING = 91;
const WHEELCHAIR = 92;
const CLIMBING = 93;
const WALKING_TREADMILL = 94;
const KICK_SCOOTER = 95;
const STAND_UP_PADDLING = 96;
const TRAIL_RUNNING = 97;
const ROWING_INDOORS = 98;
const FLOORBALL = 99;
const ICE_SKATING = 100;
const SKI_TOURING = 101;
const ROPE_JUMPING = 102;
const STRETCHING = 103;
const CANICROSS = 104;
const PADDLE_TENNIS = 105;
const PARAGLIDING = 106;

var sport = ({
    __proto__: null,
    RUNNING: RUNNING,
    CYCLING_TRANSPORT: CYCLING_TRANSPORT,
    CYCLING_SPORT: CYCLING_SPORT,
    MOUNTAIN_BIKINGS: MOUNTAIN_BIKINGS,
    SKATING: SKATING,
    ROLLER_SKIING: ROLLER_SKIING,
    SKIING_CROSS_COUNTRY: SKIING_CROSS_COUNTRY,
    SKIING_DOWNHILL: SKIING_DOWNHILL,
    SNOWBOARDING: SNOWBOARDING,
    KAYAKING: KAYAKING,
    KITE_SURFING: KITE_SURFING,
    ROWING: ROWING,
    SAILING: SAILING,
    WINDSURFING: WINDSURFING,
    FINTESS_WALKING: FINTESS_WALKING,
    GOLFING: GOLFING,
    HIKING: HIKING,
    ORIENTEERING: ORIENTEERING,
    WALKING: WALKING,
    RIDING: RIDING,
    SWIMMING: SWIMMING,
    CYCLING_INDOOR: CYCLING_INDOOR,
    OTHER: OTHER,
    AEROBICS: AEROBICS,
    BADMINTON: BADMINTON,
    BASEBALL: BASEBALL,
    BASKETBALL: BASKETBALL,
    BOXING: BOXING,
    CLIMBING_STAIRS: CLIMBING_STAIRS,
    CRICKET: CRICKET,
    ELLIPTICAL_TRAINING: ELLIPTICAL_TRAINING,
    DANCING: DANCING,
    FENCING: FENCING,
    FOOTBALL_AMERICAN: FOOTBALL_AMERICAN,
    FOOTBALL_RUGBY: FOOTBALL_RUGBY,
    FOOTBALL_SOCCER: FOOTBALL_SOCCER,
    HANDBALL: HANDBALL,
    HOCKEY: HOCKEY,
    PILATES: PILATES,
    POLO: POLO,
    SCUBA_DIVING: SCUBA_DIVING,
    SQUASH: SQUASH,
    TABLE_TENIS: TABLE_TENIS,
    TENNIS: TENNIS,
    VOLEYBALL_BEACH: VOLEYBALL_BEACH,
    VOLEYBALL_INDOOR: VOLEYBALL_INDOOR,
    WEIGHT_TRAINING: WEIGHT_TRAINING,
    YOGA: YOGA,
    MARTINAL_ARTS: MARTINAL_ARTS,
    GYMNASTICS: GYMNASTICS,
    STEP_COUNTER: STEP_COUNTER,
    CIRKUIT_TRAINING: CIRKUIT_TRAINING,
    RUNNING_TREADMILL: RUNNING_TREADMILL,
    SKATEBOARDING: SKATEBOARDING,
    SURFING: SURFING,
    SNOWSHOEING: SNOWSHOEING,
    WHEELCHAIR: WHEELCHAIR,
    CLIMBING: CLIMBING,
    WALKING_TREADMILL: WALKING_TREADMILL,
    KICK_SCOOTER: KICK_SCOOTER,
    STAND_UP_PADDLING: STAND_UP_PADDLING,
    TRAIL_RUNNING: TRAIL_RUNNING,
    ROWING_INDOORS: ROWING_INDOORS,
    FLOORBALL: FLOORBALL,
    ICE_SKATING: ICE_SKATING,
    SKI_TOURING: SKI_TOURING,
    ROPE_JUMPING: ROPE_JUMPING,
    STRETCHING: STRETCHING,
    CANICROSS: CANICROSS,
    PADDLE_TENNIS: PADDLE_TENNIS,
    PARAGLIDING: PARAGLIDING
});

const EVERYONE = 0;
const FRIENDS = 1;
const ME = 2;

var privacy = ({
    __proto__: null,
    EVERYONE: EVERYONE,
    FRIENDS: FRIENDS,
    ME: ME
});

var sportNames = {
  [RUNNING]: 'Running',
  [CYCLING_TRANSPORT]: 'Cycling, transport',
  [CYCLING_SPORT]: 'Cycling, sport',
  [MOUNTAIN_BIKINGS]: 'Mountain biking',
  [SKATING]: 'Skating',
  [ROLLER_SKIING]: 'Roller skiing',
  [SKIING_CROSS_COUNTRY]: 'Skiing, cross country',
  [SKIING_DOWNHILL]: 'Skiing, downhill',
  [SNOWBOARDING]: 'Snowboarding',
  [KAYAKING]: 'Kayaking',
  [KITE_SURFING]: 'Kite surfing',
  [ROWING]: 'Rowing',
  [SAILING]: 'Sailing',
  [WINDSURFING]: 'Windsurfing',
  [FINTESS_WALKING]: 'Fitness walking',
  [GOLFING]: 'Golfing',
  [HIKING]: 'Hiking',
  [ORIENTEERING]: 'Orienteering',
  [WALKING]: 'Walking',
  [RIDING]: 'Riding',
  [SWIMMING]: 'Swimming',
  [CYCLING_INDOOR]: 'Cycling, Indoor',
  [OTHER]: 'Other',
  [AEROBICS]: 'Aerobics',
  [BADMINTON]: 'Badminton',
  [BASEBALL]: 'Baseball',
  [BASKETBALL]: 'Basketball',
  [BOXING]: 'Boxing',
  [CLIMBING_STAIRS]: 'Climbing stairs',
  [CRICKET]: 'Cricket',
  [ELLIPTICAL_TRAINING]: 'Elliptical training',
  [DANCING]: 'Dancing',
  [FENCING]: 'Fencing',
  [FOOTBALL_AMERICAN]: 'Football, American',
  [FOOTBALL_RUGBY]: 'Football, rugby',
  [FOOTBALL_SOCCER]: 'Football, soccer',
  [HANDBALL]: 'Handball',
  [HOCKEY]: 'Hockey',
  [PILATES]: 'Pilates',
  [POLO]: 'Polo',
  [SCUBA_DIVING]: 'Scuba diving',
  [SQUASH]: 'Squash',
  [TABLE_TENIS]: 'Table tennis',
  [TENNIS]: 'Tennis',
  [VOLEYBALL_BEACH]: 'Volleyball, beach',
  [VOLEYBALL_INDOOR]: 'Volleyball, indoor',
  [WEIGHT_TRAINING]: 'Weight training',
  [YOGA]: 'Yoga',
  [MARTINAL_ARTS]: 'Martial arts',
  [GYMNASTICS]: 'Gymnastics',
  [STEP_COUNTER]: 'Step counter',
  [CIRKUIT_TRAINING]: 'Circuit Training',
  [SKATEBOARDING]: 'Skateboarding',
  [CLIMBING]: 'Climbing',
  [KICK_SCOOTER]: 'Kick scooter',
  [CANICROSS]: 'Canicross',
  [FLOORBALL]: 'Floorball',
  [ICE_SKATING]: 'Ice skating',
  [RUNNING_TREADMILL]: 'Running (Treadmill)',
  [SURFING]: 'Surfing',
  [SNOWSHOEING]: 'Showshoeing',
  [WHEELCHAIR]: 'Wheelchair',
  [WALKING_TREADMILL]: 'Walking (Treadmill)',
  [STAND_UP_PADDLING]: 'Stand up paddling',
  [TRAIL_RUNNING]: 'Trail running',
  [ROWING_INDOORS]: 'Rowing (indoors)',
  [SKI_TOURING]: 'Ski touring',
  [ROPE_JUMPING]: 'Rope jumping',
  [STRETCHING]: 'Stretching',
  [PADDLE_TENNIS]: 'Paddle tennis',
  [PARAGLIDING]: 'Paragliding'
};

const ENDOMONDO_URL = 'https://www.endomondo.com';
const ENDOMONDO_MOBILE_URL = 'https://api.mobile.endomondo.com/mobile';

export { ENDOMONDO_MOBILE_URL, ENDOMONDO_URL, privacy as PRIVACY, sport as SPORT, sportNames as SPORT_NAMES };
