// endomondo-api-handler/dist/constants/sport.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const RUNNING = 0;
const CYCLING_TRANSPORT = 1;
const CYCLING_SPORT = 2;
const MOUNTAIN_BIKINGS = 3;
const SKATING = 4;
const ROLLER_SKIING = 5;
const SKIING_CROSS_COUNTRY = 6;
const SKIING_DOWNHILL = 7;
const SNOWBOARDING = 8;
const KAYAKING = 9;
const KITE_SURFING = 10;
const ROWING = 11;
const SAILING = 12;
const WINDSURFING = 13;
const FINTESS_WALKING = 14;
const GOLFING = 15;
const HIKING = 16;
const ORIENTEERING = 17;
const WALKING = 18;
const RIDING = 19;
const SWIMMING = 20;
const CYCLING_INDOOR = 21;
const OTHER = 22;
const AEROBICS = 23;
const BADMINTON = 24;
const BASEBALL = 25;
const BASKETBALL = 26;
const BOXING = 27;
const CLIMBING_STAIRS = 28;
const CRICKET = 29;
const ELLIPTICAL_TRAINING = 30;
const DANCING = 31;
const FENCING = 32;
const FOOTBALL_AMERICAN = 33;
const FOOTBALL_RUGBY = 34;
const FOOTBALL_SOCCER = 35;
const HANDBALL = 36;
const HOCKEY = 37;
const PILATES = 38;
const POLO = 39;
const SCUBA_DIVING = 40;
const SQUASH = 41;
const TABLE_TENIS = 42;
const TENNIS = 43;
const VOLEYBALL_BEACH = 44;
const VOLEYBALL_INDOOR = 45;
const WEIGHT_TRAINING = 46;
const YOGA = 47;
const MARTINAL_ARTS = 48;
const GYMNASTICS = 49;
const STEP_COUNTER = 50;
const CIRKUIT_TRAINING = 87;
const RUNNING_TREADMILL = 88;
const SKATEBOARDING = 89;
const SURFING = 90;
const SNOWSHOEING = 91;
const WHEELCHAIR = 92;
const CLIMBING = 93;
const WALKING_TREADMILL = 94;
const KICK_SCOOTER = 95;
const STAND_UP_PADDLING = 96;
const TRAIL_RUNNING = 97;
const ROWING_INDOORS = 98;
const FLOORBALL = 99;
const ICE_SKATING = 100;
const SKI_TOURING = 101;
const ROPE_JUMPING = 102;
const STRETCHING = 103;
const CANICROSS = 104;
const PADDLE_TENNIS = 105;
const PARAGLIDING = 106;

exports.AEROBICS = AEROBICS;
exports.BADMINTON = BADMINTON;
exports.BASEBALL = BASEBALL;
exports.BASKETBALL = BASKETBALL;
exports.BOXING = BOXING;
exports.CANICROSS = CANICROSS;
exports.CIRKUIT_TRAINING = CIRKUIT_TRAINING;
exports.CLIMBING = CLIMBING;
exports.CLIMBING_STAIRS = CLIMBING_STAIRS;
exports.CRICKET = CRICKET;
exports.CYCLING_INDOOR = CYCLING_INDOOR;
exports.CYCLING_SPORT = CYCLING_SPORT;
exports.CYCLING_TRANSPORT = CYCLING_TRANSPORT;
exports.DANCING = DANCING;
exports.ELLIPTICAL_TRAINING = ELLIPTICAL_TRAINING;
exports.FENCING = FENCING;
exports.FINTESS_WALKING = FINTESS_WALKING;
exports.FLOORBALL = FLOORBALL;
exports.FOOTBALL_AMERICAN = FOOTBALL_AMERICAN;
exports.FOOTBALL_RUGBY = FOOTBALL_RUGBY;
exports.FOOTBALL_SOCCER = FOOTBALL_SOCCER;
exports.GOLFING = GOLFING;
exports.GYMNASTICS = GYMNASTICS;
exports.HANDBALL = HANDBALL;
exports.HIKING = HIKING;
exports.HOCKEY = HOCKEY;
exports.ICE_SKATING = ICE_SKATING;
exports.KAYAKING = KAYAKING;
exports.KICK_SCOOTER = KICK_SCOOTER;
exports.KITE_SURFING = KITE_SURFING;
exports.MARTINAL_ARTS = MARTINAL_ARTS;
exports.MOUNTAIN_BIKINGS = MOUNTAIN_BIKINGS;
exports.ORIENTEERING = ORIENTEERING;
exports.OTHER = OTHER;
exports.PADDLE_TENNIS = PADDLE_TENNIS;
exports.PARAGLIDING = PARAGLIDING;
exports.PILATES = PILATES;
exports.POLO = POLO;
exports.RIDING = RIDING;
exports.ROLLER_SKIING = ROLLER_SKIING;
exports.ROPE_JUMPING = ROPE_JUMPING;
exports.ROWING = ROWING;
exports.ROWING_INDOORS = ROWING_INDOORS;
exports.RUNNING = RUNNING;
exports.RUNNING_TREADMILL = RUNNING_TREADMILL;
exports.SAILING = SAILING;
exports.SCUBA_DIVING = SCUBA_DIVING;
exports.SKATEBOARDING = SKATEBOARDING;
exports.SKATING = SKATING;
exports.SKIING_CROSS_COUNTRY = SKIING_CROSS_COUNTRY;
exports.SKIING_DOWNHILL = SKIING_DOWNHILL;
exports.SKI_TOURING = SKI_TOURING;
exports.SNOWBOARDING = SNOWBOARDING;
exports.SNOWSHOEING = SNOWSHOEING;
exports.SQUASH = SQUASH;
exports.STAND_UP_PADDLING = STAND_UP_PADDLING;
exports.STEP_COUNTER = STEP_COUNTER;
exports.STRETCHING = STRETCHING;
exports.SURFING = SURFING;
exports.SWIMMING = SWIMMING;
exports.TABLE_TENIS = TABLE_TENIS;
exports.TENNIS = TENNIS;
exports.TRAIL_RUNNING = TRAIL_RUNNING;
exports.VOLEYBALL_BEACH = VOLEYBALL_BEACH;
exports.VOLEYBALL_INDOOR = VOLEYBALL_INDOOR;
exports.WALKING = WALKING;
exports.WALKING_TREADMILL = WALKING_TREADMILL;
exports.WEIGHT_TRAINING = WEIGHT_TRAINING;
exports.WHEELCHAIR = WHEELCHAIR;
exports.WINDSURFING = WINDSURFING;
exports.YOGA = YOGA;
