// endomondo-api-handler/dist/constants/sport.esm.js
const RUNNING = 0;
const CYCLING_TRANSPORT = 1;
const CYCLING_SPORT = 2;
const MOUNTAIN_BIKINGS = 3;
const SKATING = 4;
const ROLLER_SKIING = 5;
const SKIING_CROSS_COUNTRY = 6;
const SKIING_DOWNHILL = 7;
const SNOWBOARDING = 8;
const KAYAKING = 9;
const KITE_SURFING = 10;
const ROWING = 11;
const SAILING = 12;
const WINDSURFING = 13;
const FINTESS_WALKING = 14;
const GOLFING = 15;
const HIKING = 16;
const ORIENTEERING = 17;
const WALKING = 18;
const RIDING = 19;
const SWIMMING = 20;
const CYCLING_INDOOR = 21;
const OTHER = 22;
const AEROBICS = 23;
const BADMINTON = 24;
const BASEBALL = 25;
const BASKETBALL = 26;
const BOXING = 27;
const CLIMBING_STAIRS = 28;
const CRICKET = 29;
const ELLIPTICAL_TRAINING = 30;
const DANCING = 31;
const FENCING = 32;
const FOOTBALL_AMERICAN = 33;
const FOOTBALL_RUGBY = 34;
const FOOTBALL_SOCCER = 35;
const HANDBALL = 36;
const HOCKEY = 37;
const PILATES = 38;
const POLO = 39;
const SCUBA_DIVING = 40;
const SQUASH = 41;
const TABLE_TENIS = 42;
const TENNIS = 43;
const VOLEYBALL_BEACH = 44;
const VOLEYBALL_INDOOR = 45;
const WEIGHT_TRAINING = 46;
const YOGA = 47;
const MARTINAL_ARTS = 48;
const GYMNASTICS = 49;
const STEP_COUNTER = 50;
const CIRKUIT_TRAINING = 87;
const RUNNING_TREADMILL = 88;
const SKATEBOARDING = 89;
const SURFING = 90;
const SNOWSHOEING = 91;
const WHEELCHAIR = 92;
const CLIMBING = 93;
const WALKING_TREADMILL = 94;
const KICK_SCOOTER = 95;
const STAND_UP_PADDLING = 96;
const TRAIL_RUNNING = 97;
const ROWING_INDOORS = 98;
const FLOORBALL = 99;
const ICE_SKATING = 100;
const SKI_TOURING = 101;
const ROPE_JUMPING = 102;
const STRETCHING = 103;
const CANICROSS = 104;
const PADDLE_TENNIS = 105;
const PARAGLIDING = 106;

export { AEROBICS, BADMINTON, BASEBALL, BASKETBALL, BOXING, CANICROSS, CIRKUIT_TRAINING, CLIMBING, CLIMBING_STAIRS, CRICKET, CYCLING_INDOOR, CYCLING_SPORT, CYCLING_TRANSPORT, DANCING, ELLIPTICAL_TRAINING, FENCING, FINTESS_WALKING, FLOORBALL, FOOTBALL_AMERICAN, FOOTBALL_RUGBY, FOOTBALL_SOCCER, GOLFING, GYMNASTICS, HANDBALL, HIKING, HOCKEY, ICE_SKATING, KAYAKING, KICK_SCOOTER, KITE_SURFING, MARTINAL_ARTS, MOUNTAIN_BIKINGS, ORIENTEERING, OTHER, PADDLE_TENNIS, PARAGLIDING, PILATES, POLO, RIDING, ROLLER_SKIING, ROPE_JUMPING, ROWING, ROWING_INDOORS, RUNNING, RUNNING_TREADMILL, SAILING, SCUBA_DIVING, SKATEBOARDING, SKATING, SKIING_CROSS_COUNTRY, SKIING_DOWNHILL, SKI_TOURING, SNOWBOARDING, SNOWSHOEING, SQUASH, STAND_UP_PADDLING, STEP_COUNTER, STRETCHING, SURFING, SWIMMING, TABLE_TENIS, TENNIS, TRAIL_RUNNING, VOLEYBALL_BEACH, VOLEYBALL_INDOOR, WALKING, WALKING_TREADMILL, WEIGHT_TRAINING, WHEELCHAIR, WINDSURFING, YOGA };
